src/language/tsLanguageService.ts

```diff
@@ -0,0 +1,184 @@
+import { Connection } from 'vscode-languageserver/node';
+import type {
+  SignatureHelp,
+  SignatureInformation,
+  ParameterInformation,
+} from 'vscode-languageserver/node';
+import ts from 'typescript';
+import { TextDocument, Position } from 'vscode-languageserver-textdocument';
+import { readFileSync } from 'fs';
+import { join, basename, dirname } from 'path';
+
+type TypeScriptServiceHost = {
+  getLanguageService(jsDocument: TextDocument): ts.LanguageService;
+  getCompilationSettings(): ts.CompilerOptions;
+  dispose(): void;
+};
+
+// Server folder.
+const serverPath =
+  basename(__dirname) === 'dist'
+    ? dirname(__dirname)
+    : dirname(dirname(__dirname));
+
+// TypeScript library folder.
+const librarPath = join(serverPath, 'node_modules/typescript/lib');
+const contents: { [name: string]: string } = {};
+
+export default class TypeScriptService {
+  _connection: Connection;
+  _host: TypeScriptServiceHost;
+
+  constructor(connection: Connection) {
+    this._host = this._getTypeScriptServiceHost();
+    this._connection = connection;
+  }
+
+  _loadLibrary(name: string) {
+    let content = contents[name];
+    if (typeof content !== 'string' && librarPath) {
+      const libPath = join(librarPath, name); // From source.
+      try {
+        content = readFileSync(libPath).toString();
+      } catch (e) {
+        this._connection.console.error(
+          `Unable to load library ${name} at ${libPath}`
+        );
+        content = '';
+      }
+      contents[name] = content;
+    }
+    return content;
+  }
+
+  _getTypeScriptServiceHost() {
+    const compilerOptions = {
+      allowNonTsExtensions: true,
+      allowJs: true,
+      lib: ['lib.es2020.full.d.ts'], // Should match to lib from tsconfig.json.
```

Comment:
Which `tsconfig.json` are we referring to here? This is about playgrounds (… right), so there wouldn’t normally be a `tsconfig.json` for those, correct?